zkevm-circuits/src/blobkzg_circuit/src/main.rs

```diff
@@ -0,0 +1,585 @@
+        }
+
+        layouter.assign_region(
+            || "BlobKZGCircuit Region",
+            |region| {
+                // 1. FIRST ASSIGN ALL THE WITNESSES/INPUTS
+                let mut ctx = RegionCtx::new(region, 0);
+                let assigned_x = assign_wrong_integer(&integer_chip, &mut ctx, x)?;
+                ctx.next();
+                let assigned_w = integer_chip.assign_constant(&mut ctx, w)?;
+                ctx.next();
+                let assigned_zero = integer_chip.assign_constant(&mut ctx, W::ZERO)?;
+                ctx.next();
+                let assigned_one = integer_chip.assign_constant(&mut ctx, W::ONE)?;
+                ctx.next();
+                let assigned_n = integer_chip.assign_constant(&mut ctx, n)?;
+                ctx.next();
+                let assigned_datapoints: Vec<AssignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>> =
+                    datapoints
+                        .iter()
+                        .map(|d_i| {
+                            let assigned_d_i = assign_wrong_integer(&integer_chip, &mut ctx, *d_i)?;
+                            ctx.next();
+                            Ok(assigned_d_i)
+                        })
+                        .collect::<Result<Vec<_>, Error>>()?;
+                assert!(assigned_datapoints.len() == NUMBER_OF_POINTS);
+                let assigned_wis: Vec<AssignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>> = wis
+                    .iter()
+                    .map(|w_i| {
+                        let assigned_w_i = assign_wrong_integer(&integer_chip, &mut ctx, *w_i)?;
+                        ctx.next();
+                        Ok(assigned_w_i)
+                    })
+                    .collect::<Result<Vec<_>, Error>>()?;
+                assert!(assigned_wis.len() == 4096);
+                let assigned_evals: Vec<AssignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>> =
+                    evals
+                        .iter()
+                        .map(|eval_i| {
+                            let assigned_eval_i =
+                                assign_wrong_integer(&integer_chip, &mut ctx, *eval_i)?;
+                            ctx.next();
+                            Ok(assigned_eval_i)
+                        })
+                        .collect::<Result<Vec<_>, Error>>()?;
+                assert!(assigned_evals.len() == 4098);
+                let assigned_xns: [AssignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>; 257] = xns
+                    .iter()
+                    .map(|x_i| {
+                        let assigned_x_i = assign_wrong_integer(&integer_chip, &mut ctx, *x_i)?;
+                        ctx.next();
+                        Ok(assigned_x_i)
+                    })
+                    .collect::<Result<Vec<_>, Error>>()?
+                    .try_into()
+                    .unwrap();
+                let assigned_x2is: [AssignedInteger<W, N, NUMBER_OF_LIMBS, BIT_LEN_LIMB>; 256] =
```

Comment:
Question: do we need to allocate the entire `x2is` array? We use the powers only for the bits that aren't enabled